### DISCUSSION TITLE
Magic link Emails, Verification emails Fail Randomly and Throw Errors

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#2740] [Q&A]
I didn't want to post an issue about this, as I'm not sure if I'm missing a setting or not. But, in our app Password less/Magiclink sign in emails, Verification emails and other emails that require some sort of ticket, just randomly fail out of nowhere, with this error: 
``js
error=invalid-ticket&errorDescription=Invalid+or+expired+verification+ticket
``

This happens even if the email is just sent and so it not actually related to any expiration. We can't debug this really because all of these errors are caused after the redirect from the nhost server auth URL, so we have no idea why they just start failing and the user is sent into some loop of constant failed verifications.  After some time, everything starts working again, but by then the user has given up. We are not sure what causes this or how to fix it. We have isolated a few test cases where this happens repeatedly as described below. We were thinking that maybe your server has some sort of setting that doesn't allow multiple verification attempts within a certain timeframe and has some reset? Otherwise we can't understand why these emails fail randomly and then start to work again after some time. 

**Scenario 1:** 
1. new user signs up and gets sent a verification email. 
2. Do not click on that email to simulate a situation where a user does not get the email for whatever reason (this is common scenario b/c sometimes an email goes to spam or a user is impatient).
3. Resend another verification email to the user within a minute or so.
4. If you click on the 2nd verification email, you get the error above. Subsequent attempts at new verifications all fail. Reset password emails fail and magic Link emails fail.

**Scenario 2:**
1. User signs up and verifies email.
2. User forgets password for some reason (again somewhat common)
3. User sends password reset email. Verification fails 
4. User sends magic link. Verification fails. 

FYI: We already use the suggested protection for magic links as documented in this post: https://nhost.io/blog/protect-magic-links-from-email-clients, so we know it's not an email client issue. And per above, the links start working again after some time, sometimes 10 minutes, sometimes less. hard to isolate.

**Comment by osseonews** [comment]
Also, in attempting to debug the problems above, we started looking into the solution we implemented following this post, https://nhost.io/blog/protect-magic-links-from-email-clients, and I'm not sure I entirely understand how the solution proposed there actually solves the problem. Because what happens is that the email client (e.g. Gmail), opens the email and visits the new verify link set up in the app, triggering the useEffect and then the redirect to nhost. Then when the actual user gets the email and clicks on the link nhost throws an error because the ticket was already used by the email client. So how exactly does creating the separate page in an app actually help? It just substitutes one link for another.  The email client still visits the link no matter where it is, invalidating the ticket by the time the user actually clicks on the link. I'm tempted to believe this is what is causing our issues because I cannot figure out any other reason why these links just fail randomly.

**Comment by osseonews** [comment]
So after extensive testing, I can confirm that the error above happens due to the way various email clients process the verify link using the Protect Magic Link strategy per the blog. I recommended changing the code to something like the code below,  to introduce some sort of delay tactic to prevent execution of the useEffect by the email client, which triggers the redirect and invalidates the ticket for the actual user. 

Of course, I do not think this is a "proper" solution, as this is just becoming a game of cat & mouse with the email clients and actually alot of enterprise systems won't even allow a click to any link. The proper solution is to do away with links and use an OTP-like code instead for everything. You can easily implement this now by just replacing your uuidv4() generation of the code with a 6 digit code, which would allow backwards compatibility to anyone still using links and at the same time allow those who want to move away from links to present a proper OTP-like form to users based on accepted UI/UX for OTP.  I posted about this solution in the issue related to this. Thanks.

```js

const VerifyPage: React.FC = () => {
  const nhost = useNhostClient();
  const [loading, setLoading] = useState(true);
  const location = useLocation();
  const currentSearchParams = location.search;
  const searchParams = new URLSearchParams(currentSearchParams.toString());
  
  useEffect(() => {
    const ticket = searchParams.get("ticket");
    const redirectTo = searchParams.get("redirectTo");
    const type = searchParams.get("type");

    if (ticket && redirectTo && type && !loading) {
    window.location.href = `${nhost.auth.url}/verify?ticket=${ticket}&type=${type}&redirectTo=${redirectTo}`;
    }
    setTimeout(() => {
      setLoading(false);
    }, 1500); // Delay for 1.5 second setting loading to prevent unnecessary redirects. 1.5 seconds seems to work consistently for now
  }, [searchParams, nhost?.auth?.url, loading]);

  if (loading) {
    return (
      <div className="flex items-center justify-center min-h-screen">
        <div className="-mx-4 flex w-full justify-center">
          
            Loading
            
        </div>
      </div>
    );
  }
  
  return (
    <div className="flex items-center justify-center min-h-screen">
      <div className="-mx-4 flex w-full justify-center">
        
          Verifying
         
      </div>
    </div>
  );
};

export default VerifyPage;

```

**Comment by osseonews** [comment]
After thinking about this a bit more, I think  a better solution might be to replace the useEffect entirely with an onSubmit handler with a form. The email clients dont submit the form from what i can tell. This requires an extra action from the user, but this seems better than having these links expire all the time or to have to delay the page load to prevent unwanted execution of the useEffect. Example code below. 

```js
const VerifyPage: React.FC = () => {
  const nhost = useNhostClient();
  const [loading, setLoading] = useState(true);
  const location = useLocation();
  const currentSearchParams = location.search;
  const searchParams = new URLSearchParams(currentSearchParams.toString());
  const handleSubmit = async (event) => {
    event.preventDefault();
    setLoading(false);
    try {
    const ticket = searchParams.get("ticket");
    const redirectTo = searchParams.get("redirectTo");
    const type = searchParams.get("type");
 	    if (ticket && redirectTo && type && !loading) {
	    window.location.href = `${nhost.auth.url}/verify?ticket=${ticket}&type=${type}&redirectTo=${redirectTo}`;

	    } else {
//throw error           
}
    } catch (error) {
        // Handle error - set state
    }
    
  };
 
  if (loading) {
    return (
      <div className="flex items-center justify-center min-h-screen">
        <div className="-mx-4 flex w-full justify-center">
          
          Thanks for verifying your email. To finalize your sign in, please use the button below.
          
          <form onSubmit={handleSubmit}>
            {/* Your form fields here */}
            <button type="submit">Sign Me In Now</button>
          </form>
        </div>
      </div>
    );
  }
  
  return (
    <div className="flex items-center justify-center min-h-screen">
      <div className="-mx-4 flex w-full justify-center">
        
          Signing in
          
       
      </div>
    </div>
  );
};

export default VerifyPage;
```

**Comment by dbarrosop** [comment]
> After thinking about this a bit more, I think a better solution might be to replace the useEffect entirely with an onSubmit handler with a form

That's what the example should be doing as, otherwise, as you pointed out it will be triggered by scanner. I will ask a colleague to review and confirm how the example works and update if necessary. Apologies if the example was misleading.

**Reply by dbarrosop** [reply]
I talked internally with my colleague and I can confirm the example does what you suggested. I just want to reiterate that's not 100% correct as some crawlers might implement/run javascript. For this to be effective this must require some manual action, like pushing a button or similar.

We will update the example accordingly. Thanks for pointing this out.

**Reply by osseonews** [reply]
thanks for verifying. btw, I also didn't use a simple button and instead put it in a form, as I think that none of the email client crawlers will actually submit a form. But, I haven't tested it really to see if a basic button without a form would also work - just seemed that adding a form would provide an extra layer of protection against fake submissions.

**Reply by dbarrosop** [reply]
To track it: https://github.com/nhost/nhost/discussions/2740

Thanks again
